### PR TITLE
[Layers] Added RotaryEmbedding forward for bc mode & Fixed rope ut

### PR DESCRIPTION
**Pull request by abenmao**

### CODE REVIEW

File: src/kernels/rotary_embedding_kernels.cpp [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+//     k_embed = (k * cos) + (rotate_half(k) * sin)
+//     return q_embed, k_embed
+//
+
+void llamaApplyRotaryPosEmbed(float *query, float *key, float *emb_cos, float *emb_sin, int qStride, int kStride,
+        int dim, int totSeqLen, int qHeads, int kHeads, const int *positionIds) {
+    const int half = (dim + 1) / 2;
+    const int heads = std::max(qHeads, kHeads);
+
+#pragma omp parallel for collapse(2)
```
Comment (pujiang2018) [inline src/kernels/rotary_embedding_kernels.cpp:59]: currently we have imbalance tasks inside omp block, we could optimize it in next stage.

File: src/kernels/rotary_embedding_kernels.cpp [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+                __mmask16 mask = (remain >= 16 ? 0xffff : (1 << remain) - 1);
+
+                __m512 pCosVec = _mm512_maskz_loadu_ps(mask, &pcos[i]);
+                __m512 pSinVec = _mm512_maskz_loadu_ps(mask, &psin[i]);
+
+                // Compute something like:
+                // q[i] = q[i] * pcos[i] - q[i + half] * psin[i];
+                // q[i + half] = q[i + half] * pcos[i + half] + q[i] * psin[i + half];
+                if (head < qHeads) {
+                    __m512 qVec = bfloat16_t::cvt_bf16_to_fp32(_mm256_maskz_loadu_epi16(mask, &q[i]));
```
Comment (pujiang2018) [inline src/kernels/rotary_embedding_kernels.cpp:113]: now we can use xft::load_avx512 to make it shorter.

File: src/layers/attention.h [hunk 3]
```diff
@@ -468,22 +468,23 @@ class Attention {
...
-        //         std::iota(posIds.begin(), posIds.end(), pastSeqLen);
-        //     }
-        //     qkpo.forward(query.Data(), key.Data(), query.Stride(), key.Stride(), qkShape, posIds.data());
-        // }
+        if (ctx->maxPosEmbed > 0) {
+            int qheads = this->endQHead - this->startQHead;
+            int kheads = this->endKVHead - this->startKVHead;
+            int totInputSeqLen = 0;
+            for (auto seq : seqs) {
+                totInputSeqLen += seq->getInputSeqLen();
```
Comment (pujiang2018) [inline src/layers/attention.h:476]: will put totInputSeqLen into parameters in next stage.

File: src/layers/yarn_scaled_rotary_embedding.cpp [hunk 4]
```diff
@@ -160,22 +113,22 @@ void LlamaYaRNScaledRotaryEmbedding::forward(
         for (int bs = 0; bs < batchSize; ++bs) {
             for (int seq = 0; seq < seqLen; ++seq) {
                 int pos = positionIds[seq];
-                float *pcos = embCos + pos * dim;
-                float *psin = embSin + pos * dim;
+                float *pcos = embCos + pos * half;
```
Comment (pujiang2018) [inline src/layers/yarn_scaled_rotary_embedding.cpp:116]: Original code is:
float *pcos = embCos + pos * dim;
are we sure need to change to current code?
float *pcos = embCos + pos * half;

Comment (abenmao) [reply src/layers/yarn_scaled_rotary_embedding.cpp:116]: Previously, the lens of pcos and psin were “dim”, with the first half and the second half being same. Now, pcos and psin have been unified to inv_freq_size (dim / 2), so we made the change here.

File: src/layers/attention.h (hunk 3 above)
Comment (abenmao) [reply src/layers/attention.h:476]: OK~ Then we will remove these line here.

File: src/kernels/rotary_embedding_kernels.cpp (hunk 2 above)
Comment (abenmao) [reply src/kernels/rotary_embedding_kernels.cpp]: Done~

File: src/kernels/rotary_embedding_kernels.cpp (hunk 1 above)
Comment (abenmao) [reply src/kernels/rotary_embedding_kernels.cpp:60]: Each row of q and k can be parallel, so it should be balanced~

Comment (pujiang2018) [reply src/kernels/rotary_embedding_kernels.cpp:60]: when kHeads < qHeads, it may cause imbalance issue. but anyhow, next step.

**Pull request merged**